plugins/nominal-connection-checker/src/type_hierarchy.js

```diff
@@ -439,4 +563,140 @@ class TypeDef {
   hasDescendant(descendantName) {
     return this.descendants_.has(descendantName);
   }
+
+  /**
+   * Returns an array of this type's parameters, in the order for its superType.
+   * @param {string} ancestorName The caseless name of the ancestor to get the
+   *     parameters for.
+   * @param {!Array<!TypeStructure>=} actualTypes Optional actual types to
+   *     substitute for parameters. These types may be generic.
+   * @return {!Array<!TypeStructure>} This type's parameters, in the order for
+   *     its superType.
+   */
+  getParamsForAncestor(ancestorName, actualTypes = undefined) {
+    if (ancestorName == this.name && !this.paramsMap_.has(this.name)) {
+      // Convert this type's params to a type structure.
+      this.paramsMap_.set(
+          this.name,
+          this.params_.map((param) => {
+            return {name: param.name, params: []};
+          }));
+    }
+    const params = this.paramsMap_.get(ancestorName);
```

Comment:
Is this getting a copy, or are you directly modifying the `paramsMap` with the following `replaceFn`?

Reply:
Yeah this should be doing a copy, fixed :D In the next PR (where I move and expand the type structure stuff) I'll change it to do a deep copy as well.